src/extension/csv-with-metadata-header/src/main/java/au/org/emii/geoserver/wfs/response/CsvPivotedFeatureCollectionSource.java

```diff
@@ -0,0 +1,195 @@
+import org.geotools.feature.visitor.UniqueVisitor;
+import org.geotools.jdbc.JDBCDataStore;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+import org.opengis.feature.simple.SimpleFeature;
+import org.opengis.feature.simple.SimpleFeatureType;
+import org.opengis.feature.type.AttributeDescriptor;
+import org.postgresql.util.PGobject;
+
+import javax.servlet.ServletContext;
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.*;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+import static au.org.emii.geoserver.wfs.response.CSVWithMetadataHeaderOutputFormat.JSON_FIELD_TYPE;
+import static org.geotools.gce.imagemosaic.Utils.FF;
+
+
+public class CsvPivotedFeatureCollectionSource implements CsvSource {
+
+    private static final int MAX_PIVOT_COLUMNS = 10000;
+    private final SimpleFeatureIterator featureIterator;
+    private final ArrayList<String> nonPivotColumnNames;
+    private final List<String> pivotColumnNames;
+    private final List<String> pivotSourceColumnNames;
+    private final PivotConfig pivotConfig;
+
+    static Logger LOGGER = org.geotools.util.logging.Logging.getLogger("au.org.emii.geoserver.wfs.response");
+
+    private SimpleFeature nextFeature;
+    private final Catalog catalog;
+    private final ServletContext context;
+
+    public CsvPivotedFeatureCollectionSource(Operation getFeatureOperation, SimpleFeatureCollection featureCollection, Catalog catalog, ServletContext context, JDBCDataStore dataStore) {
+        this.featureIterator = featureCollection.features();
+        this.catalog = catalog;
+        this.context = context;
+
+        // Retrieve config params stored in pivot.xml
+
+        PivotConfigFile pivotConfigurationFile = new PivotConfigFile(getDataDirectory(getFeatureOperation));
+        this.pivotConfig = pivotConfigurationFile.getConfig();
+
+        //  Get feature type definition
+        
+        SimpleFeatureType ft = featureCollection.getSchema();
+
+        this.nonPivotColumnNames = new ArrayList<>();
+        pivotSourceColumnNames = new ArrayList<>();
+
+        // Determine pivoted (JSON) and non-pivoted columns
+
+        for (AttributeDescriptor descriptor: ft.getAttributeDescriptors()) {
+            String attributeName = descriptor.getLocalName();
+            String attributeType = descriptor.getUserData().size() > 0 ? descriptor.getUserData().get("org.geotools.jdbc.nativeTypeName").toString() : "";
+
+            if (attributeType.equals(JSON_FIELD_TYPE)) {
+                pivotSourceColumnNames.add(attributeName);
+            } else if (!pivotConfig.getExcludedFields().contains(attributeName)) {
+                this.nonPivotColumnNames.add(attributeName);
+            }
+        }
+
+        // Read the first row to get JSON attr for column names
```

Comment:
No need to do this any more?

Reply:
Yes and no, the first row needs to be initialised but the column names do not need to be retrieved from it.